File: assets/lib/src/screens/designprinciples/composite/hierarchy_composite.dart
import 'dart:io';

// Component Interface
abstract class FileSystemEntity {
  void showDetails([int indent = 0]);
}

// Leaf Class - File
class FileEntity implements FileSystemEntity {
  String name;

  FileEntity(this.name);

  @override
  void showDetails([int indent = 0]) {
    print('${' ' * indent}- $name');
  }
}

// Composite Class - Folder
class Folder implements FileSystemEntity {
  String name;
  List<FileSystemEntity> children = [];

  Folder(this.name);

  void add(FileSystemEntity entity) {
    children.add(entity);
  }

  @override
  void showDetails([int indent = 0]) {
    print('${' ' * indent}[Folder] $name');
    for (var child in children) {
      child.showDetails(indent + 2);
    }
  }
}

// Main function to test the Composite Pattern
void main() {
  var root = Folder('Root');
  var subFolder = Folder('SubFolder');

  root.add(FileEntity('file1.txt'));
  root.add(subFolder);
  subFolder.add(FileEntity('file2.txt'));
  subFolder.add(FileEntity('file3.txt'));

  root.showDetails();
}

File: assets/lib/src/screens/designprinciples/composite/hierarchy_composite.cpp
#include <iostream>
#include <vector>
#include <memory>

// Component Interface
class FileSystemEntity {
public:
    virtual void showDetails(int indent = 0) const = 0;
    virtual ~FileSystemEntity() = default;
};

// Leaf Class - File
class File : public FileSystemEntity {
    std::string name;
public:
    File(const std::string& name) : name(name) {}
    void showDetails(int indent = 0) const override {
        std::cout << std::string(indent, ' ') << "- " << name << "\n";
    }
};

// Composite Class - Folder
class Folder : public FileSystemEntity {
    std::string name;
    std::vector<std::shared_ptr<FileSystemEntity>> children;
public:
    Folder(const std::string& name) : name(name) {}

    void add(const std::shared_ptr<FileSystemEntity>& entity) {
        children.push_back(entity);
    }

    void showDetails(int indent = 0) const override {
        std::cout << std::string(indent, ' ') << "[Folder] " << name << "\n";
        for (const auto& child : children) {
            child->showDetails(indent + 2);
        }
    }
};

int main() {
    auto root = std::make_shared<Folder>("Root");
    auto subFolder = std::make_shared<Folder>("SubFolder");

    root->add(std::make_shared<File>("file1.txt"));
    root->add(subFolder);
    subFolder->add(std::make_shared<File>("file2.txt"));
    subFolder->add(std::make_shared<File>("file3.txt"));

    root->showDetails();
    return 0;
}
